distribution/lib/Standard/Base/0.0.0-dev/src/Error/Common.enso

```diff
@@ -33,3 +183,379 @@ type Wrapped_Dataflow_Error payload
+    catch : Any -> Any -> (Caught_Panic -> Any) -> Any
+    catch panic_type ~action handler =
+        Panic.catch_primitive action caught_panic->
+            case Meta.get_polyglot_language panic_type == "java" of
+                False -> case caught_panic.payload.is_a panic_type of
+                    True -> handler caught_panic
+                    False -> Panic.throw caught_panic
+                True -> case caught_panic.payload of
+                    Polyglot_Error java_exception ->
+                        case Java.is_instance java_exception panic_type of
+                            True -> handler caught_panic
+                            False -> Panic.throw caught_panic
+                    _ -> Panic.throw caught_panic
+
+    ## Executes the provided action and converts a possible panic matching any of
+       the provided types into a dataflow Error.
+
+       If action executes successfully, the result of `Panic.recover` is the result
+       of that action. Otherwise, if it panicked with a type matching one of the
+       expected error types, that panic is returned as a dataflow error. Unexpected
+       panics are passed through as-is. it is the panic that was thrown after
+       conversion to a dataflow error.
+
+       Arguments:
+       - expected_types: The types of expected panics which should be recovered.
+         This can either be a Vector of types or a single type.
+       - action: The code to execute that potentially panics.
+
+       > Example
+         Converting an expected panic to a dataflow error.
+
+             Panic.recover Illegal_Argument_Error (Panic.throw (Illegal_Argument_Error "Oh!" Nothing))
+
+       > Example
+         Converting one of many expected panic types to a dataflow error.
+
+             Panic.recover [Illegal_Argument_Error, Illegal_State_Error] (Panic.throw (Illegal_Argument_Error "Oh!" Nothing))
+    recover : (Vector.Vector Any | Any) -> Any -> Any
+    recover expected_types ~action =
+        types_to_check = case expected_types of
+            Vector.Vector _ -> expected_types
+            _ -> [expected_types]
+        Panic.catch Any action caught_panic->
+            is_matched = types_to_check.exists typ->
+                caught_panic.payload.is_a typ
+            case is_matched of
+                True -> caught_panic.convert_to_dataflow_error
+                False -> Panic.throw caught_panic
+
+## The runtime representation of a syntax error.
+
+   Arguments:
+   - message: A description of the erroneous syntax.
+@Builtin_Type
+type Syntax_Error message
+
+## The runtime representation of a type error.
+
+   Arguments:
+   - expected: The expected type at the error location.
+   - actual: The actual type at the error location.
+   - name: The name of the argument whose type is mismatched.
+@Builtin_Type
+type Type_Error expected actual name
```

Comment:
This file is pretty big, maybe it would be better to have separate 3 files for:
- `Error` type
- `Caught_Panic` and `Panic` types
- the base error types like these ones.

I think this way it would be easier to look around the codebase, as files will be smaller and more topical.